[ASCII-2435] Hostname resolution for IMDSv2 transition

louis-cqrl opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

This PR aims to add `legacy-resolution-hostname` to metadata payload, which will contain the hostname fetch without IMDSv2 scenario.

It also adds `ec2_imdsv2_transition`, when this config flag is set, IMDSv2 is used by default, if IMDSv2 call fails then it falls back on DMI method to fetch the instance ID.
It also send `legacy-resolution-hostname` in the hostname metadata payload if it is different from the hostname

### Motivation

Some dashboards could be impacted by IMDSv2 transition, this PR aims to minimise this impact by saving the actual hostname with a new field if the hostname is changing

### Describe how to test/QA your changes
Here are the following tests performed:

#### IMDSv2 and IMDSv1 enabled on the EC2 instance:
```
ec2-hostname: ip-10-1-60-231.ec2.internal
host_aliases: [i-0a19aed46c9e3abe5]
hostname: i-0a19aed46c9e3abe5
hostname-resolution-version: 1
instance-id: i-0a19aed46c9e3abe5
socket-fqdn: ip-10-1-60-231.ec2.internal.
socket-hostname: ip-10-1-60-231
hostname provider: aws
```
The output is similar to the current release.


Set `ec2_imdsv2_transition_payload_enabled` at `true`. The output is the same. This is expected as `  legacy-resolution-hostname` is the same as `hostname`


#### IMDSv2 enabled and IMDSv1 disabled on the EC2 instance:
```
 host_aliases: [i-023b337a43a03eeb6]
  hostname: ip-10-1-61-76
  hostname-resolution-version: 1
  socket-fqdn: ip-10-1-61-76.ec2.internal.
  socket-hostname: ip-10-1-61-76
  hostname provider: os
```
The output is similar to the current release.

`ec2_imdsv2_transition_payload_enabled`: true
```
  host_aliases: [i-023b337a43a03eeb6]
  hostname: i-023b337a43a03eeb6
  hostname-resolution-version: 1
  instance-id: i-023b337a43a03eeb6
  legacy-resolution-hostname: ip-10-1-61-76
  socket-fqdn: ip-10-1-61-76.ec2.internal.
  socket-hostname: ip-10-1-61-76
  hostname provider: aws
  unused hostname providers:
```
The hostname is an instance id and  `legacy-resolution-hostname` is the old hostname

`ec2_imdsv2_transition_payload_enabled: true` and `ec2_prefer_imdsv2: true`  
```
  host_aliases: [i-023b337a43a03eeb6]
  hostname: i-023b337a43a03eeb6
  hostname-resolution-version: 1
  instance-id: i-023b337a43a03eeb6
  socket-fqdn: ip-10-1-61-76.ec2.internal.
  socket-hostname: ip-10-1-61-76
  hostname provider: aws
```
`legacy-resolution-hostname` is not sent but this is expected.

Testing DMI: The endpoint `ec2_imdsv2_transition_payload_enabled` was disabled. `hostname` was the os hostname (`ip-`). 
When `legacy-resolution-hostname` is set at `True`, `hostname` was the instance id (`i-`) because it used DMI.

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

louis-cqrl committed:
feat(comp/metadata/host/hostimpl/utils/meta.go): Add LegacyResolutionHostname field to `Meta` struct as an empty string

louis-cqrl committed:
Reworking of `forceIMDSv2` boolean to an IMDSv2 action enum

ogaca-dd commented on pkg/util/hostname/providers.go:
```diff
@@ -185,6 +206,20 @@ func GetWithProvider(ctx context.Context) (Data, error) {
 }
 
+// GetWithProviderWithoutCacheAndIMDSV2 returns the hostname for the Agent and the provider that was use to retrieve it without using the cache and without using IMDSv2
+func GetWithProviderWithoutCacheAndIMDSV2(ctx context.Context) (Data, error) {
```
First, I think it makes sense to use cache for the new field. Also having a cache or not is implementation specific and shouldn't be reflected in the naming of the function when a cache is not used. What do you think?

louis-cqrl committed:
Move LegacyResolutionHostname to the same location as the hostname

fix(comp/metadata/host/hostimpl/utils/meta.go): Move LegacyResolutionHostname logic to GetPayload function
feat(comp/metadata/host/hostimpl/utils/host.go): Add LegacyResolutionHostname field to HostMetadata if different from the hostname
feat(comp/collector/collector/collectorimpl/agent_check_metadata.go): Add LegacyResolutionHostname field to AgentCheckMetadata if different from the hostname

louis-cqrl committed:
fix(pkg/util/ec2/imds_helpers.go): use assertion for IMDSv2 action instead of inequality

louis-cqrl committed:
Rename `fromEC2WithIMDSV2` to `fromEC2`

louis-cqrl committed:
fix(pkg/util/hostname/providers.go): change ec2 provider attribution to be clearer

louis-cqrl committed:
fix(pkg/util/hostname/providers.go): Move `GetWithProvider` function code above

louis-cqrl committed:
Create special `GetInstanceID` for non IMDSv2 action

louis-cqrl committed:
Cache Legacy resolution hostname value

louis-cqrl committed:
Rename `disableIMDSv2` to `notUseIMDSv2`

louis-cqrl committed:
fix(pkg/util/hostname/providers.go): remove disableIMDSv2 boolean when not needed

louis-cqrl committed:
fix(pkg/util/ec2/ec2.go): add fetcher for no IMDSv2 instanceID

louis-cqrl committed:
Update pkg/util/hostname/providers.go

Co-authored-by: Olivier G <52180542+ogaca-dd@users.noreply.github.com>

louis-cqrl committed:
Rename IMDSv2Action enum to IMDSv2Usage

louis-cqrl committed:
Rollback on boolean for getHostname method

louis-cqrl commented on pkg/util/hostname/providers.go:
```diff
@@ -150,7 +181,14 @@ func GetWithProvider(ctx context.Context) (Data, error) {
+		notUseIMDSv2 = true
+	} else {
+		notUseIMDSv2 = false
+	}
```
I added a boolean to the getHostname method previously but wanted to avoid to change constructors to much, but you are right let's go back to the boolean it will be safer

louis-cqrl committed:
Add `host_meta_version` to meta payload

louis-cqrl committed:
Use IMDSv2 by default when `ec2_imdsv2_transition` is set

louis-cqrl committed:
Add DMI fallback for the ec2 provider when `ec2_imdsv2_transition` is set

vickenty commented on comp/collector/collector/collectorimpl/agent_check_metadata.go:
```diff
@@ -55,12 +55,18 @@ func (p *Payload) SplitPayload(_ int) ([]marshaler.AbstractMarshaler, error) {
+		meta.HostMetaVersion = "1.0"
+	}
+
+	cp := hostMetadataUtils.GetCommonPayload(hostnameData, c.config)
```
Since this logic is shared between this and host payload, both of which include CommonPayload, would it make sense to report this data as part of the common payload? This way this IMDSv2 logic would need to be implemented only once.

louis-cqrl replied:
done

hush-hush commented on pkg/util/ec2/ec2.go:
```diff
@@ -96,14 +96,27 @@ func GetSourceName() string {
 	Attempt: func(ctx context.Context) (interface{}, error) {
-		return getMetadataItemWithMaxLength(ctx, imdsInstanceID, false)
+		hostname, err := getMetadataItemWithMaxLength(ctx, imdsInstanceID, UseIMDSv2(false, false))
+		if err == nil && hostname != "" {
```
We should not have to check both the error and the value (ie: the error should not be nil if we consider an empty hostname an error).

louis-cqrl replied:
done

louis-cqrl committed:
turn `GetMeta` method to private

hush-hush commented on pkg/util/ec2/imds_helpers.go:
```diff
@@ -57,39 +66,50 @@ func getMetadataItemWithMaxLength(ctx context.Context, endpoint string, forceIMD
 // UseIMDSv2 returns true if the agent should use IMDSv2
-func UseIMDSv2(forceIMDSv2 bool) bool {
-	return pkgconfigsetup.Datadog().GetBool("ec2_prefer_imdsv2") || forceIMDSv2
+func UseIMDSv2(force bool, disable bool) IMDSv2Usage {
```
It's not intuitive to have 2 params to force using or not using V2.

What the point of calling `UseIMDSv2(true, false)` or `UseIMDSv2(false, true)` ? could we not simply replace those call by `true` and `false` and have this function only be use for case where we want to check the configuration ?

louis-cqrl replied:
> It's not intuitive to have 2 params to force using or not using V2.

I agree to this

> What the point of calling UseIMDSv2(true, false) or UseIMDSv2(false, true) ?

The idea was mainly to avoid the duplication of boolean to each function prototype, passing (..., force bool, disable bool) to each function it goes through was a bit heavy.

>could we not simply replace those call by true and false and have this function only be use for case where we want to check the configuration ?

Each time I call the function is because we want to check the configuration, thought about replacing directly with the IMDSv2 action to take but it needs to check the config to know if they need to use IMDSv2 or not.

Nonetheless I agree that's iffy, a solution may be to duplicate or refactor the code logic to achieve it in a more elegant way however it may need more times to do it.

WDYT ?

louis-cqrl replied:
done

sgnn7 replied:
+1 on this function needing a bit of clean-up

louis-cqrl replied:
done in https://github.com/DataDog/datadog-agent/pull/30760

hush-hush commented on pkg/util/ec2/imds_helpers.go:
```diff
@@ -24,6 +24,15 @@ var (
+
+const (
+	notUseIMDSv2 IMDSv2Usage = iota
+	useIMDSv2
+	forceIMDSv2
```
What the difference between `useIMDSv2` and `forceIMDSv2` ? When would we have `useIMDSv2` not use IMDV2 ?

Is `forceIMDSv2` for when we only want to use IMDSv2 ? If so I'd rename this to (or something similar):
- `useIMDSv1`
- `useIMDSv2Andv1`
- `useIMDSv2`

louis-cqrl replied:
`forceIMDSv2` enforce the usage of v2 and if it fails it should not fallback on v1.
However we can rename them as you described

louis-cqrl replied:
done

louis-cqrl committed:
Move `LegacyResolutionHostname` logic to `getMeta` function

hush-hush commented on pkg/util/hostname/providers.go:
```diff
@@ -139,7 +151,22 @@ func saveHostname(cacheHostnameKey string, hostname string, providerName string)
+
+// GetWithProviderWithoutIMDSV2 returns the hostname for the Agent and the provider that was use to retrieve it without using IMDSv2
+func GetWithProviderWithoutIMDSV2(ctx context.Context) (Data, error) {
+	// If the user has set the ec2_prefer_imdsv2 then IMDSv2 is used by default by the user, `legacy_resolution_hostname` is not needed for the transition
+	// If the user has set the ec2_imdsv2_transition then IMDSv2 is used by default by the agent, `legacy_resolution_hostname` is needed for the transition
+	if pkgconfigsetup.Datadog().GetBool("ec2_prefer_imdsv2") || !pkgconfigsetup.Datadog().GetBool("ec2_imdsv2_transition") {
+		return Data{}, nil
+	}
```
I think this should be done on the caller side. If we call a method named `GetWithProviderWithoutIMDSV2`, it should return the hostname without IMDSv2. WDYT ?

louis-cqrl replied:
changed the implementation to try to focus the ec2 logic inside the ec2 package

hush-hush commented on pkg/util/hostname/providers.go:
```diff
@@ -139,7 +151,22 @@ func saveHostname(cacheHostnameKey string, hostname string, providerName string)
+	if pkgconfigsetup.Datadog().GetBool("ec2_prefer_imdsv2") || !pkgconfigsetup.Datadog().GetBool("ec2_imdsv2_transition") {
+		return Data{}, nil
+	}
+	data, err := getHostname(ctx, "legacy_resolution_hostname", true)
+	return data, err
```
```suggestion
	return getHostname(ctx, "legacy_resolution_hostname", true)
```

hush-hush commented on pkg/util/hostname/providers.go:
```diff
@@ -62,64 +64,74 @@ type provider struct {
-		expvarName:       "os",
-	},
-	{
+func getProviderCatalog(notUseIMDSv2 bool) []provider {
+	providerCatalog := []provider{
```
Would you mind keep the providers static in a var block and create with 2 independant AWS providers and then have the method choose which one to use ?

Something like:
```
var (
	configProvider = provider{
			name:             configProvider,
			cb:               fromConfig,
			stopIfSuccessful: true,
			expvarName:       "'hostname' configuration/environment",
	},
        [...]
)

func getProviderCatalog(withoutIMDSv2 bool) []provider {
	providers := []{
			configProvider,
			hostnameFileProvider,
			[...]
	}

	if withoutIMDSv2 (
		providers = append(providers, awsNoIMDSV2Provider)
	} else {
		providers = append(providers, awsIMDSV1Provider)
	}
	return providers 
}
```

louis-cqrl replied:
done

louis-cqrl commented on comp/metadata/host/hostimpl/utils/meta.go:
```diff
@@ -25,15 +25,17 @@ var (
+	AgentHostname            string   `json:"agent-hostname,omitempty"`
+	ClusterName              string   `json:"cluster-name,omitempty"`
+	LegacyResolutionHostname string   `json:"legacy-resolution-hostname,omitempty"`
+	HostMetaVersion          string   `json:"host-meta-version,omitempty"`
```
sound like a good name to me, done

louis-cqrl commented on comp/collector/collector/collectorimpl/agent_check_metadata.go:
```diff
@@ -55,12 +55,18 @@ func (p *Payload) SplitPayload(_ int) ([]marshaler.AbstractMarshaler, error) {
+	legacyResolutionHostnameData, _ := hostname.GetWithProviderWithoutIMDSV2(ctx)
+	if legacyResolutionHostnameData.Hostname != meta.Hostname {
+		meta.LegacyResolutionHostname = legacyResolutionHostnameData.Hostname
+		meta.HostMetaVersion = "1.0"
```
I think this is due to backend implementation, nonetheless I'll let @ogaca-dd answer to this in order to have a precise answer

sgnn7 replied:
> I think we should have a number for all cases.
I'd probably agree with that unless there's a good reason not to do that

> why not fill LegacyResolutionHostname in all cases even if it's the same as Hostname ?

@hush-hush It's an optimization so that we don't send a useless field to the backend.

louis-cqrl committed:
Rename `host_meta_version` to `hostname-resolution-version`

louis-cqrl committed:
Update pkg/util/hostname/providers.go

Co-authored-by: maxime mouial <hush-hush@users.noreply.github.com>

louis-cqrl committed:
put providers as static variable

louis-cqrl committed:
re-order provider

louis-cqrl committed:
fix(pkg/util/ec2.g): avoid checking both the error and the value

louis-cqrl committed:
rename notUseIMDSv2 to legacyHostnameResolution

louis-cqrl committed:
linters

louis-cqrl committed:
Rename IMDSv2Usage fields and replace value directly when already known

louis-cqrl committed:
linters

sgnn7 commented on pkg/util/ec2/imds_helpers.go:
```diff
@@ -57,39 +66,50 @@ func getMetadataItemWithMaxLength(ctx context.Context, endpoint string, forceIMD
 	source := metadataSourceIMDSv1
 	headers := map[string]string{}
-	if UseIMDSv2(forceIMDSv2) {
+	if imdsv2Usage == useIMDSv2Andv1 || imdsv2Usage == useIMDSv2 {
```
I think this logic may need a bit of clarification (maybe just naming to clarify things). As it stands, it's very unclear what `imdsv2Usage` stands for without scrolling up to see what the enum contains. Maybe something like:
```go
if allowedIMDSVersions == IMDSAllVersions || allowedIMDSVersions == IMDSVersion2 {
```

Alternatively, maybe this is something that can be improved with a real object but this may be too much to do on short notice:
```go
func doHTTPRequest(ctx context.Context, url string, ec2IMDSVersionConfig EC2IMDSVersionConfig) (string, error) {
...
if ec2IMDSVersionConfig.enablesV2() {
...
  if ec2IMDSVersionConfig.forcesV2() {
  ...
```

louis-cqrl replied:
renaming done with some small changes

louis-cqrl replied:
I will do something like that:

```
func doHTTPRequest(ctx context.Context, url string, ec2IMDSVersionConfig EC2IMDSVersionConfig) (string, error) {
...
if ec2IMDSVersionConfig.enablesV2() {
...
  if ec2IMDSVersionConfig.forcesV2() {
  ...
```

As you mentioned in a future PR (after the test PR being done) since it may brings some behaviour changes if it works for you ?

louis-cqrl replied:
done in: [`9fdf71d` (#30760)](https://github.com/DataDog/datadog-agent/pull/30760/commits/9fdf71d23d34c820a57b04c02fa20e4c74a558f5)

sgnn7 commented on pkg/util/ec2/network.go:
```diff
@@ -18,7 +18,7 @@ import (
 	Name: "EC2 Public IPv4 Address",
 	Attempt: func(ctx context.Context) (interface{}, error) {
-		return getMetadataItem(ctx, imdsIPv4, false)
+		return getMetadataItem(ctx, imdsIPv4, UseIMDSv2(false, false))
```
We'll need some renaming likely. Having `UseIMDSv2()` and `useIMDSv2` in the same file but being completely different seems like a developer foot-gun.

louis-cqrl replied:
I renamed the enum it''s fixed

sgnn7 reviewed (commented):
@louis-cqrl I didn't have enough time to cover all the bases so I'll defer to others on the logic flow but I left a few comments (mostly around naming/style).

louis-cqrl commented on pkg/config/setup/config.go:
```diff
@@ -560,6 +560,7 @@ func InitConfig(config pkgconfigmodel.Setup) {
 	config.BindEnvAndSetDefault("collect_ec2_tags", false)
 	config.BindEnvAndSetDefault("collect_ec2_tags_use_imds", false)
 	config.BindEnvAndSetDefault("exclude_ec2_tags", []string{})
+	config.BindEnvAndSetDefault("ec2_imdsv2_transition", false)
```
done

louis-cqrl committed:
rename ec2_imdsv2_transition to ec2_imdsv2_transition_payload_enabled

louis-cqrl committed:
Update pkg/util/ec2/ec2.go

Co-authored-by: Srdjan Grubor <sgnn7@sgnn7.org>

louis-cqrl committed:
Merge branch 'louis-cqrl/IMDSV2-add-LegacyResolutionHostname-field' of github.com:DataDog/datadog-agent into louis-cqrl/IMDSV2-add-LegacyResolutionHostname-field

louis-cqrl committed:
fix error log

louis-cqrl committed:
Update pkg/util/hostname/providers.go

Co-authored-by: Srdjan Grubor <sgnn7@sgnn7.org>

louis-cqrl committed:
Merge branch 'louis-cqrl/IMDSV2-add-LegacyResolutionHostname-field' of github.com:DataDog/datadog-agent into louis-cqrl/IMDSV2-add-LegacyResolutionHostname-field

hush-hush commented on pkg/util/hostname/providers.go:
```diff
@@ -62,64 +62,99 @@ type provider struct {
-	},
+	}
+
+	ec2HostnameResolutionProvider = provider{
```
or `ec2HostnameLegacyProvider` ?

louis-cqrl replied:
done in https://github.com/DataDog/datadog-agent/pull/30760

hush-hush commented on pkg/util/hostname/providers.go:
```diff
@@ -62,64 +62,99 @@ type provider struct {
+		osProvider,
+	}
+
+	if legacyHostnameResolution {
+		providerCatalog = append(providerCatalog, ec2HostnameResolutionProvider)
+	} else {
+		providerCatalog = append(providerCatalog, ec2Provider)
+	}
```
Feel like the code is easier to follow with a single return and a simple if/else :man_shrugging:

louis-cqrl replied:
I agree with @hush-hush I would prefer to keep the ec2 logic for providers inside 1 if/else block

louis-cqrl commented on pkg/util/ec2/ec2.go:
```diff
@@ -96,14 +96,26 @@ func GetSourceName() string {
 	Attempt: func(ctx context.Context) (interface{}, error) {
-		return getMetadataItemWithMaxLength(ctx, imdsInstanceID, false)
+		hostname, err := getMetadataItemWithMaxLength(ctx, imdsInstanceID, UseIMDSv2(false, false))
+		if err != nil && pkgconfigsetup.Datadog().GetBool("ec2_imdsv2_transition") {
```
> can we put the flag name in a const as well?

done

> I think we want to see the value of the bool config flag in a separate debug/trace log statement before this condition otherwise we won't know if an error made us skip pulling the DMI value or the lack of an error when debugging.

The error seems to be logged here: https://github.com/DataDog/datadog-agent/blob/3f972c1d3b36d56bc34a8732483644f95363041b/pkg/util/hostname/providers.go#L205

I just split the if condition in 2 if conditions to only have the DMI log when the flag is set to true, I will let you check to see if it fits to you

louis-cqrl committed:
create const var for ec2_imdsv2_transition_payload_enabled

louis-cqrl committed:
Use int for HostnameResolutionVersion

louis-cqrl commented on comp/metadata/host/hostimpl/utils/meta.go:
```diff
@@ -63,13 +65,19 @@ func GetMeta(ctx context.Context, conf model.Reader) *Meta {
+		HostAliases:               cloudproviders.GetHostAliases(ctx),
+		InstanceID:                instanceID,
+		AgentHostname:             agentHostname,
+		HostnameResolutionVersion: "1.0",
```
done

louis-cqrl commented on comp/metadata/host/hostimpl/utils/meta.go:
```diff
@@ -63,13 +65,19 @@ func GetMeta(ctx context.Context, conf model.Reader) *Meta {
+		HostnameResolutionVersion: "1.0",
+	}
+
+	legacyResolutionHostnameData, _ := hostname.GetWithProviderWithLegacyHostnameResolution(ctx)
```
I agree, I applied this change,
other proposition: `GetWithProviders` becomes a parent function that calls either the legacy resolution or the other one, it would make some refactoring so if we want to do this maybe keep it for an other PR ?

louis-cqrl committed:
rename GetWithProviderWithLegacyHostnameResolution function to GetWithProviderLegacyResolution

louis-cqrl committed:
rename GetInstanceIDWithLegacyHostnameResolution to GetLegacyResolutionInstanceID

louis-cqrl commented on pkg/util/hostname/common.go:
```diff
@@ -25,14 +25,15 @@ import (
-	osHostnameUsable  = isOSHostnameUsable
+	isFargateInstance                            = fargate.IsFargateInstance
+	ec2GetInstanceID                             = ec2.GetInstanceID
+	ec2GetInstanceIDWithLegacyHostnameResolution = ec2.GetInstanceIDWithLegacyHostnameResolution
```
done

louis-cqrl committed:
move comment around for hostname provider catalog

louis-cqrl committed:
Renaming of imdsUsage enum

ogaca-dd commented on pkg/util/ec2/imds_helpers.go:
```diff
@@ -24,13 +24,13 @@ var (
-// IMDSv2Usage is an enum to determine how to interact with the IMDSv2 option
-type IMDSv2Usage int
+// IMDSVersion is an enum to determine how to interact with the IMDSv2 option
+type IMDSVersion int
```
`IMDSVersion` doesn't need to be exported.

louis-cqrl replied:
good catch, done

ogaca-dd reviewed (approved):
LGTM

louis-cqrl committed:
fix(ec2.go): update comment

louis-cqrl committed:
Update pkg/util/ec2/ec2.go

Co-authored-by: Olivier G <52180542+ogaca-dd@users.noreply.github.com>

louis-cqrl committed:
rename `UseIMDSv2` to `GetIMDSVersion`

louis-cqrl committed:
Merge branch 'louis-cqrl/IMDSV2-add-LegacyResolutionHostname-field' of github.com:DataDog/datadog-agent into louis-cqrl/IMDSV2-add-LegacyResolutionHostname-field

louis-cqrl committed:
remove unneeded export of IMDSVersion

louis-cqrl commented on pkg/util/ec2/ec2.go:
```diff
@@ -112,14 +124,24 @@ func GetInstanceID(ctx context.Context) (string, error) {
 	return instanceIDFetcher.FetchString(ctx)
 }
 
+// GetInstanceIDWithLegacyHostnameResolution fetches the instance id for current host from the EC2 metadata API without using IMDSv2 and DMI
```
done

louis-cqrl committed:
fix: linter

louis-cqrl committed:
fix(pkg/util/hostname/providers.go) avoid updating the legacy resolution hostname providers

hush-hush reviewed (approved):
Added a few last nit-pick that we can solve in later PRs.

louis-cqrl commented on pkg/util/ec2/imds_helpers.go:
```diff
@@ -57,39 +66,50 @@ func getMetadataItemWithMaxLength(ctx context.Context, endpoint string, forceIMD
-func UseIMDSv2(forceIMDSv2 bool) bool {
-	return pkgconfigsetup.Datadog().GetBool("ec2_prefer_imdsv2") || forceIMDSv2
+// getIMDSVersion returns true if the agent should use IMDSv2
+func getIMDSVersion(force bool, disable bool) imdsVersion {
```
Yes I think this should be reworked, for this PR, I'll just keep the renaming, and do the refactor in an other PR

louis-cqrl replied:
done in https://github.com/DataDog/datadog-agent/pull/30760

louis-cqrl committed:
fix(pkg/util/ec2/ec2.go, pkg/util/hostname/providers.go): Move config check to avoid running providers if ec2_imdsv2_transition_payload_enabled is not set

louis-cqrl committed:
fix(pkg/util/ec2/ec2.go, pkg/util/ec2/imds_helpers.go): avoid to update metadata source if we are retrieving the legacy resolution hostname
fix(pkg/util/ec2/ec2_tags.go, pkg/util/ec2/network.go): update call to function with the boolean parameter to update the metadata source

ogaca-dd reviewed (approved):
LGTM

louis-cqrl commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

louis-cqrl commented:
/merge

louis-cqrl commented:
/merge
